writer/separate_tree_test.go

```diff
@@ -0,0 +1,205 @@
+package writer
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"strings"
+	"testing"
+
+	"github.com/dineshba/tf-summarize/terraformstate"
+	"github.com/stretchr/testify/assert"
+
+	. "github.com/hashicorp/terraform-json"
+	tfjson "github.com/hashicorp/terraform-json"
+)
+
+func TestSeparateTree_Write_DrawableTrue(t *testing.T) {
+
+	changes := map[string]terraformstate.ResourceChanges{
+		"add": {
+			{
+				Address: "aws_instance.example1",
+				Change:  &Change{Actions: Actions{ActionCreate}},
+			},
+		},
+		"delete": {
+			{
+				Address: "aws_instance.example2",
+				Change:  &Change{Actions: Actions{ActionDelete}},
+			},
+		},
+	}
+
+	tw := NewSeparateTree(changes, true)
+	var buf bytes.Buffer
+	err := tw.Write(&buf)
+
+	assert.NoError(t, err)
+	expectedOutput := `################### ADD ###################
+      ╭─╮      
+      │.│      
+      ╰┬╯      
+       │       
+╭──────┴─────╮ 
+│aws_instance│ 
+╰──────┬─────╯ 
+       │       
+ ╭─────┴─────╮ 
+ │example1(+)│ 
+ ╰───────────╯ 
+
+
+################### DELETE ###################
+      ╭─╮      
+      │.│      
+      ╰┬╯      
+       │       
+╭──────┴─────╮ 
+│aws_instance│ 
+╰──────┬─────╯ 
+       │       
+ ╭─────┴─────╮ 
+ │example2(-)│ 
+ ╰───────────╯ 
+
+
+`
+
+	assert.Equal(t, expectedOutput, removeANSI(buf.String()))
+}
+
+func TestSeparateTree_Write_NonDrawable(t *testing.T) {
+
+	changes := map[string]terraformstate.ResourceChanges{
+		"add": {
+			{
+				Address: "aws_instance.example1",
+				Change:  &Change{Actions: Actions{ActionCreate}},
+			},
+		},
+		"delete": {
+			{
+				Address: "aws_instance.example2",
+				Change:  &Change{Actions: Actions{ActionDelete}},
+			},
+		},
+	}
+
+	tw := NewSeparateTree(changes, false)
+	var buf bytes.Buffer
+	err := tw.Write(&buf)
+
+	assert.NoError(t, err)
+
+	expectedOutput := `################### ADD ###################
+|---aws_instance
+|	|---example1(+)
+
+
+################### DELETE ###################
+|---aws_instance
+|	|---example2(-)
+
+
+`
+
+	assert.Equal(t, expectedOutput, removeANSI(buf.String()))
+}
+
+// Mock writer that returns an error after a certain number of writes
+type controlledErrorWriter struct {
```

Comment:
Is it possible to use gomock for this?

Reply:
Yes, it's a good idea, Have updated to use gomock to mock the io writer interface and handle the expected calls of the writer func

![image](https://github.com/user-attachments/assets/a1113704-13a5-4a16-bc77-eb70729b01f8)